evm-contracts/src/v0.7/dev/StalenessFlaggingValidator.sol

```diff
@@ -0,0 +1,149 @@
+
+  event FlagsAddressUpdated(
+    address indexed previous,
+    address indexed current
+  );
+  event FlaggingThresholdUpdated(
+    address indexed aggregator,
+    uint256 indexed previous,
+    uint256 indexed current
+  );
+
+  /**
+   * @notice Create a new StalenessFlaggingValidator
+   * @param flagsAddress Address of the flag contract
+   * @dev Ensure that this contract has sufficient write permissions
+   * on the flag contract
+   */
+  constructor(address flagsAddress) 
+    public 
+    ConfirmedOwner(msg.sender)
+  {
+    setFlagsAddress(flagsAddress);
+  }
+
+  /**
+   * @notice Updates the flagging contract address for raising flags
+   * @param flagsAddress sets the address of the flags contract
+   */
+  function setFlagsAddress(address flagsAddress)
+    public
+    onlyOwner()
+  {
+    address previous = address(s_flags);
+    if (previous != flagsAddress) {
+      s_flags = FlagsInterface(flagsAddress);
+      emit FlagsAddressUpdated(previous, flagsAddress);
+    }
+  }
+
+  /**
+   * @notice Set the threshold limits for each aggregator
+   * @dev parameters must be same length
+   * @param aggregators address[] memory
+   * @param flaggingThresholds uint256[] memory
+   */
+  function setThresholds(address[] memory aggregators, uint256[] memory flaggingThresholds)
+    public 
+    onlyOwner()
+  {
+    require(aggregators.length == flaggingThresholds.length, "Different sized arrays");
+    for (uint256 i = 0; i < aggregators.length; i++) {
+      address aggregator = aggregators[i];
+      uint256 previousThreshold = s_thresholds[aggregator];
+      uint newThreshold = flaggingThresholds[i];
+      if (previousThreshold != newThreshold) {
+        s_thresholds[aggregator] = newThreshold;
+        emit FlaggingThresholdUpdated(aggregator, previousThreshold, newThreshold);
+      }
+    }
+  }
+
+  /**
+   * @notice Check for staleness in an array of aggregators
+   * @dev If any of the aggregators are stale, this function will return true,
+   * otherwise false
+   * @param aggregators address[] memory
+   * @return bool
+   */
+  function check(address[] memory aggregators) public view returns (bool) {
+    uint256 currentTimestamp = block.timestamp;
+
+    for (uint256 i = 0; i < aggregators.length; i++) {
+      if (isStale(aggregators[i], currentTimestamp)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * @notice Check for staleness in an array of aggregators, raise a flag
+   * on the flags contract for each aggregator that is stale
+   * @dev This contract must have write permissions on the flags contract
+   * @param aggregators address[] calldata
+   */
+  function update(address[] calldata aggregators) external {
+    uint256 currentTimestamp = block.timestamp;
+
+    for (uint256 i = 0; i < aggregators.length; i++) {
+      address aggregator = aggregators[i];
+      if (isStale(aggregator, currentTimestamp)) {
+        s_flags.raiseFlag(aggregator);
```

Comment:
It would probably save a lot of gas to build up this list locally and then do a single call with the list to `raiseFlags`.

Reply:
This is tough as arrays can only be dynamic in storage. `push()` isn't available for `memory` arrays.

Reply:
We can do it, but it'll be a little messy:

1. Create a second array of equal size to the parameter called `staleAggregators`
2. Loop through aggregators, adding stale ones to `staleAggregators`, whilst keeping a count
3. Send a slice of the `staleAggregators` to `raiseFlags`, the size of which depends on the count.

Reply:
Turns out, [slices are only implemented on `calldata` arrays](https://docs.soliditylang.org/en/v0.7.4/types.html?highlight=slices#array-slices). Since we're acting on an array in memory, this becomes even messier.

I think we're going to have to live with the GAS implications of individual `raiseFlag` calls.